src/neuroglancer/datasource/brainmaps/backend.ts

```diff
@@ -207,6 +211,100 @@ export class MeshSource extends ParameterizedMeshSource<MeshSourceParameters> {
     return makeRequest(parameters['instance'], httpCall, cancellationToken)
         .then(response => decodeFragmentChunk(chunk, response));
   }
+
+  filterFragments(fragmentIds: string[]|null, clipBounds: Bounds) {
+    clipBounds;
+    if (!fragmentIds) {
+      return fragmentIds;
+    }
+
+    let filteredFragments = [];
+    for (let fragmentId of fragmentIds) {
+      // TODO(blakely): Hardcoded for now, remove when we can filter on the backend.
+      const fragmentSize = 500;
+      let fragmentBounds = this.getFragmentBounds(fragmentId, vec3.clone([fragmentSize, fragmentSize, fragmentSize]));  
+      if (this.boundsIntersect(fragmentBounds, clipBounds)) {
+        filteredFragments.push(fragmentId);
+      }      
+    }
+    
+    return filteredFragments;
+  }
+
+  private numBitsInPositiveValue(value: number) {
+    if (value < 0) {
+      throw new Error(`Undefined number of bits for value ${value}`);
+    }
+    if (value === 0) {
+      return 0;
+    }
+    value -= 1;
+    let numBits = 1;
+    while((value = value >> 1) > 0) {
+      ++numBits;
+    }
+    return numBits;
+  }
+
+  private getFragmentBounds(fragmentId: string, fragmentSize: vec3): Bounds{
+    let corner = this.getFragmentCorner(fragmentId, fragmentSize);
+
+    let halfSize = vec3.create();
+    vec3.scale(halfSize, fragmentSize, 0.5);
+    let center = vec3.create();
+    vec3.add(center, corner, halfSize);
+
+    return {
+      center,
+      size: fragmentSize,
+    };
+  }
+
+  private getFragmentCorner(fragmentId: string, fragmentSize: vec3) {
+    let id = new Uint64();
+    if (!id.tryParseString(fragmentId, 16)) {
+      throw new Error(`Couldn't parse fragmentId ${fragmentId} as hex-encoded Uint64`);
+    }
+    if (id.high) {
+      throw new Error(`Fragment ids > 2^32 not supported yet`);
+    }
+    const maxCoordBits = fragmentSize.map(size => this.numBitsInPositiveValue(size));
```

Comment:
I think the objective of this code is to decode the z-index.  Should the max bits be based on the size of the volume (which is not known here) rather than the size of the fragment?

I think there are ways to do this with just a small number of bit operations, e.g. as shown here: https://fgiesen.wordpress.com/2009/12/13/decoding-morton-codes/
That way you can avoid the looping.